File: lua/grade-school/grade-school.lua
local School = {}
School.__index = School

function School:new()
    local school = {students = {}}
    setmetatable(school, self)
    return school
end

function School:add(name, grade)
    if self.students[name] then
        return false
    end
    self.students[name] = grade
    return true
end

function School:roster()
    local names = {}
    for name, _ in pairs(self.students) do
        names[#names + 1] = name
    end
    table.sort(names, function(a, b)
        local aGrade, bGrade = self.students[a], self.students[b]
        return aGrade < bGrade or (aGrade == bGrade and a < b)
    end)
    return names
end

function School:grade(wanted)
    local names = {}
    for name, grade in pairs(self.students) do
        if grade == wanted then
            names[#names + 1] = name
        end
    end
    table.sort(names)
    return names
end

return School

File: php/grade-school/GradeSchool.php
<?php

declare(strict_types=1);

class School
{
    private $students = [];

    public function add(string $name, int $grade): void
    {
        if ($this->contains($name)) return;
        array_push($this->students, ['name' => $name, 'grade' => $grade]);
    }

    public function contains(string $name): bool
    {
        return in_array($name, array_column($this->students, 'name'));
    }

    public function grade($grade)
    {
        $class = array_filter($this->students, fn($student) => $student['grade'] == $grade);
        $names = array_column($class, 'name');
        sort($names);
        return $names;
    }

    public function studentsByGradeAlphabetical(): array
    {
        $grades = array_unique(array_column($this->students, 'grade'));
        sort($grades);
        return array_reduce(
            $grades,
            fn($roster, $grade) => $roster + [$grade => $this->grade($grade)],
            []
        );
    }
}
